### PR TITLE
added an MCIP about sorted transaction outputs

### PR DESCRIPTION
**Pull request by integral-llc**
[Rendered Text](https://github.com/mobilecoinfoundation/mcips/blob/87387c079d651b040ffc02a95d8cf4cb3cd6b2c7/text/0034-sorted-transaction-outputs.md)

### CODE REVIEW

File: text/0034-sorted-transaction-outputs.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+- Feature Name: sorted_transaction_outputs
+- Start Date: 2022-04-04
+- MCIP PR: [mobilecoinfoundation/mcips#0000](https://github.com/mobilecoinfoundation/mcips/pull/0000)
```
Comment (jcape) [inline text/0034-sorted-transaction-outputs.md:3]: This should also get updated to #34

**Review by jcape — COMMENTED** [review]
Could you include a "rendered text" link on the comment and update the MCIP PR link as well?

File: text/0034-sorted-transaction-outputs.md (hunk 1 above)
Comment (integral-llc) [reply text/0034-sorted-transaction-outputs.md]: fixed

File: text/0034-sorted-transaction-outputs.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+- Feature Name: sorted_transaction_outputs
+- Start Date: 2022-04-04
+- MCIP PR: [mobilecoinfoundation/mcips#34](https://github.com/mobilecoinfoundation/mcips/pull/34)
+- Tracking Issue: original PR: https://github.com/mobilecoinfoundation/mobilecoin/pull/561
```
Comment (remoun) [inline text/0034-sorted-transaction-outputs.md:4]: Let's file an issue referencing that PR, and link to that here.

**Review by remoun — APPROVED** [review]

File: text/0034-sorted-transaction-outputs.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+# Summary
+[summary]: #summary
+
+For blocks with a version >= 3, all the transaction's outputs will be sorted by the public key value.
```
Comment (jcape) [inline text/0034-sorted-transaction-outputs.md:9]:
```suggestion
For blocks with a version >= 3, all the transaction's outputs MUST be sorted by the public key value.
```

File: text/0034-sorted-transaction-outputs.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+* signature
+* transaction fee
+* uniqueness of key images
+* uniqueness of outputs public keys
+* tombstone
+
+With block version >= 3 the validator will also check whether the transaction outputs are sorted. 
+
+When library is compiled with `test-only` feature flag, there is a possibility to provide a custom trait for the transaction
+builder to be used for sorting the transaction outputs.
```
Comment (jcape) [inline text/0034-sorted-transaction-outputs.md:43]: This should discuss the API changes when building with the `test-only` feature to show how to produce a transaction with an incorrect order.

File: text/0034-sorted-transaction-outputs.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are no drawbacks.
```
Comment (jcape) [inline text/0034-sorted-transaction-outputs.md:48]:
```suggestion
Any third-party software which was not previously sorting outputs will need to be updated. We are not aware of any such software, however.
```

File: text/0034-sorted-transaction-outputs.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+# Summary
+[summary]: #summary
+
+For blocks with a version >= 3, all the transaction's outputs MUST be sorted by the public key value.
+
+# Motivation
+[motivation]: #motivation
+
+The main idea behind this feature is to prevent establishing the order of outputs that were added to the 
+transaction, thus making hard or impossible to identify the source.
```
Comment (nick-mobilecoin) [inline text/0034-sorted-transaction-outputs.md:15]: nit: I'm pretty bad at grammar so might want someone else to thumbs up, but my mind wants to put an `it` in 
> thus making *it* hard or impossible ...

```suggestion
transaction, thus making it hard or impossible to identify the source.
```

File: text/0034-sorted-transaction-outputs.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+* membership proof
+* signature
+* transaction fee
+* uniqueness of key images
+* uniqueness of outputs public keys
+* tombstone
+
+With block version >= 3 the validator will also check whether the transaction outputs are sorted. 
+
+When library is compiled with `test-only` feature flag, there is a possibility to provide a custom trait for the transaction
```
Comment (nick-mobilecoin) [inline text/0034-sorted-transaction-outputs.md:42]: nit: again could be my bad grammar, my mind is wanting a `the`s in 
>When *the* library is compiled with *the* `test-only` feature flag

```suggestion
When the library is compiled with the `test-only` feature flag, there is a possibility to provide a custom trait for the transaction
```

File: text/0034-sorted-transaction-outputs.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The enclave transaction validation will now check if the TxOuts are sorted.
+It will return a transaction validation error if not.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Before block version 3, the transaction validator was checking on the following checks
```
Comment (nick-mobilecoin) [inline text/0034-sorted-transaction-outputs.md:26]: nit: this is fairly subjective, so ignore as you see fit.  
The only reason I mention it is that it makes the statement more concise, and *I* don't think it loses any meaning.

```suggestion
Before block version 3, the transaction validator was checking:
```

**Review by nick-mobilecoin — APPROVED** [review]

File: text/0034-sorted-transaction-outputs.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+// use the custom Ordering implementation to generate wrongly sorted outputs
+// when building the transaction
+transaction_builder.build_with_sorter::<R, InverseTxOutputsOrdering>(rng).unwrap()
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Any third-party software which was not previously sorting outputs will need to be updated. We are not aware of any such software, however.
```
Comment (garbageslam) [inline text/0034-sorted-transaction-outputs.md:73]: i think there are a few people who are not using our rust transaction builder -- but we can loop them in on this

**Pull request merged**